File: packages/sdk-ts/src/resources/api-keys.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { ApiKey, CreateApiKey, CreatedApiKey, PaginationQuery } from '../types.js';

const enc = encodeURIComponent;

/** API keys: scoped credentials. Secrets are returned once, at creation and rotation. */
export class ApiKeys {
    constructor(private readonly http: Http) {}

    list(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<ApiKey>> {
        return this.http.list<ApiKey>('/v1/api-keys', query, opts);
    }

    /** The `key` in the result is shown once and cannot be retrieved again. A key cannot grant scopes the calling key lacks. */
    create(body: CreateApiKey, opts?: RequestOptions): Promise<CreatedApiKey> {
        return this.http.request<CreatedApiKey>('POST', '/v1/api-keys', { ...opts, body });
    }

    /** Immediate, with no grace period. */
    revoke(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/api-keys/${enc(id)}`, opts);
    }

    /** A replacement is issued now; the old key keeps working for 24 hours. */
    rotate(id: string, opts?: RequestOptions): Promise<CreatedApiKey> {
        return this.http.request<CreatedApiKey>('POST', `/v1/api-keys/${enc(id)}/rotate`, opts);
    }
}

File: packages/sdk-dart/lib/src/resources/api_keys.dart
import '../models/api_key.dart';
import '../page.dart';
import '../transport.dart';

/// API keys: scoped credentials. Secrets are returned once.
class ApiKeys {
  /// Binds the resource to a transport.
  const ApiKeys(this._transport);

  final Transport _transport;

  /// Lists keys. Secrets are never returned, only a prefix and last four.
  Future<Page<ApiKey>> list(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/api-keys', ApiKey.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Mints a key. The `key` in the result is shown once and cannot be
  /// retrieved again. A key cannot grant scopes the calling key lacks.
  Future<CreatedApiKey> create(CreateApiKey body) =>
      _transport.request('POST', '/v1/api-keys', CreatedApiKey.fromJson,
          body: body.toJson());

  /// Revokes a key. Immediate, with no grace period.
  Future<void> revoke(String id) => _transport.requestVoid(
      'DELETE', '/v1/api-keys/${Uri.encodeComponent(id)}');

  /// Issues a replacement now; the old key keeps working for 24 hours.
  Future<CreatedApiKey> rotate(String id) => _transport.request('POST',
      '/v1/api-keys/${Uri.encodeComponent(id)}/rotate', CreatedApiKey.fromJson);
}
